### PR TITLE
Orderby ability

### PR DESCRIPTION
**Pull request by valinurovam**
Hi all.
Im tried to add orderby ability for yii2-redis. It works for me. Review please.

### CODE REVIEW

File: ActiveQuery.php [hunk 1]
```diff
@@ -369,6 +365,19 @@ protected function executeScript($db, $type, $columnName = null)
      */
     private function findByPk($db, $type, $columnName = null)
     {
+        if (!empty($this->orderBy) and in_array($type, ['All', 'One', 'Count', 'Column'])) {
```
Comment (samdark) [inline ActiveQuery.php]: `and` → `&&`

File: ActiveQuery.php [hunk 2]
```diff
@@ -369,6 +365,19 @@ protected function executeScript($db, $type, $columnName = null)
...
+            $orderBy = array_slice($this->orderBy, 0);
```
Comment (samdark) [inline ActiveQuery.php]: Isn't `array_slice($this->orderBy, 0)` gives us a copy of the same array? Can't it be simple `=`?

File: ActiveQuery.php [hunk 3]
```diff
@@ -369,6 +365,19 @@ protected function executeScript($db, $type, $columnName = null)
...
+            $k = key($orderBy);
+            $v = $orderBy[$k];
+            if (is_numeric($k)) {
+                $orderColumn = $v;
+                $orderType = SORT_ASC; // ASC sort
```
Comment (samdark) [inline ActiveQuery.php]: No need for this comment, it's obvious from constant name.

File: ActiveQuery.php [hunk 4]
```diff
@@ -405,7 +418,21 @@ private function findByPk($db, $type, $columnName = null)
...
+                asort($orderArray);
+            } else {
+                arsort($orderArray);
+            }
+            foreach ($orderArray as $orderKey => $orderItem) {
+                $resultData[] = $data[$orderKey];
+            }
+        }
+
+        if (!empty($resultData))
```
Comment (samdark) [inline ActiveQuery.php]: Need `{` even for 1 line statements.

File: LuaScriptBuilder.php [hunk 5]
```diff
@@ -167,8 +163,32 @@ private function build($query, $buildResult, $return)
             $loadColumnValues .= "local $alias=redis.call('HGET',$key .. ':a:' .. pk, '$column')\n";
         }
 
+        $sortAbility = "";
+        if (!empty($query->orderBy)) {
+            $orderBy = array_slice($query->orderBy, 0);
```
Comment (samdark) [inline LuaScriptBuilder.php]: Can't it be done with simple `=`?

**Comment by cebe** [comment]
really cool, will check it in detail later. 

Could you add some unit tests? At least these two functions should be removed, which will add tests for some sorting:
https://github.com/yiisoft/yii2-redis/blob/master/tests/ActiveRecordTest.php#L146-L154

File: LuaScriptBuilder.php (hunk 5 above)
Comment (samdark) [inline LuaScriptBuilder.php]: Also, why can't we use `$query->orderBy` directly?

File: ActiveQuery.php (hunk 2 above)
Comment (samdark) [inline ActiveQuery.php]: Can we use `$this->orderBy` directly w/o copying it?

**Comment by valinurovam** [comment]
@cebe, I will try to add unit tests for the week

**Comment by valinurovam** [comment]
Find some bug... all values will be sort like strings. For example  we have numeric pk id (1,2,3,4...11,12...21,22). The result for by id will be
1,11,12,...,2,21,22,...3,31,32, but not 1,2,3,4,5...31,32,33.

Don't have any idea to solve this. May be test is_numeric for values and use type casting for numeric values

**Comment by valinurovam** [comment]
Hi all again. Tried to fix all bugs and errors. Remove duplicated test consisted orderBy call.
Can you review it again?

Tell me pls, should I add test for this case:
Customer::find()->andFilterWhere(['id' => [3,4,1]])->orderBy(['id' => SORT_DESC])->indexBy('id')->all();

File: ActiveQuery.php [hunk 6]
```diff
@@ -369,6 +366,21 @@ protected function executeScript($db, $type, $columnName = null)
      */
     private function findByPk($db, $type, $columnName = null)
     {
+        if (!empty($this->orderBy) && in_array($type, ['All', 'One', 'Count', 'Column'])) {
```
Comment (andruha) [inline ActiveQuery.php]: Not need sorting for type 'Count'.
And this expression must be assigned to a variable because it's duplicated 2 times

Comment (valinurovam) [inline ActiveQuery.php]: Thx, fixed

File: ActiveQuery.php [hunk 7]
```diff
@@ -369,6 +366,22 @@ protected function executeScript($db, $type, $columnName = null)
      */
     private function findByPk($db, $type, $columnName = null)
     {
+        $typesForOrder = ['All', 'One', 'Column'];
+        if (!empty($this->orderBy) && in_array($type, $typesForOrder)) {
```
Comment (andruha) [inline ActiveQuery.php]:
``` php
$needSort = !empty($this->orderBy) && in_array($type, ['All', 'One', 'Column']);
if ($needSort) {
...
}
```

File: ActiveQuery.php [hunk 8]
```diff
@@ -393,10 +406,14 @@ private function findByPk($db, $type, $columnName = null)
         }
         $i = 0;
         $data = [];
+        $orderArray = [];
         foreach ($pks as $pk) {
             if (++$i > $start && ($limit === null || $i <= $start + $limit)) {
                 $key = $modelClass::keyPrefix() . ':a:' . $modelClass::buildKey($pk);
                 $result = $db->executeCommand('HGETALL', [$key]);
+                if (!empty($this->orderBy) and in_array($type, $typesForOrder)) {
```
Comment (andruha) [inline ActiveQuery.php]:
``` php
if ($needSort) {
...
}
```

File: ActiveQuery.php [hunk 9]
```diff
@@ -405,7 +422,22 @@ private function findByPk($db, $type, $columnName = null)
                 }
             }
         }
-        // TODO support orderBy
+
+        $resultData = [];
+        if (!empty($this->orderBy) and in_array($type, $typesForOrder)) {
```
Comment (andruha) [inline ActiveQuery.php]:
``` php
if ($needSort) {
...
}
```

**Comment by valinurovam** [comment]
Hi all,
Can you review it again?

**Comment by valinurovam** [comment]
Any changes?

File: ActiveQuery.php [hunk 10]
```diff
@@ -369,6 +366,22 @@ protected function executeScript($db, $type, $columnName = null)
      */
     private function findByPk($db, $type, $columnName = null)
     {
+        $needSort = !empty($this->orderBy) && in_array($type, ['All', 'One', 'Column']);
+        if ($needSort) {
+            if (count($this->orderBy) > 1)
```
Comment (samdark) [inline ActiveQuery.php]: Need `{` and `}` even for one liners.

File: ActiveQuery.php [hunk 11]
```diff
@@ -369,6 +366,22 @@ protected function executeScript($db, $type, $columnName = null)
...
+                throw new NotSupportedException('orderBy by multi columns is currently not supported by redis ActiveRecord.');
```
Comment (samdark) [inline ActiveQuery.php]: multi → multiple
is currently not → is not currently

File: LuaScriptBuilder.php [hunk 12]
```diff
@@ -167,8 +163,33 @@ private function build($query, $buildResult, $return)
             $loadColumnValues .= "local $alias=redis.call('HGET',$key .. ':a:' .. pk, '$column')\n";
         }
 
-        return <<<EOF
+        $getAllPks = <<<EOF
 local allpks=redis.call('LRANGE',$key,0,-1)
+EOF;
+        if (!empty($query->orderBy)) {
+            if (count($query->orderBy) > 1)
+                throw new NotSupportedException('orderBy by multi columns is currently not supported by redis ActiveRecord.');
```
Comment (samdark) [inline LuaScriptBuilder.php]: multi → multiple

Comment (samdark) [inline LuaScriptBuilder.php]: is currently not → is not currently

**Comment by koxu1996** [comment]
Hey, what about merging it?

**Comment by nnrudakov** [comment]
Excuse me, have repo team plan to merge this pr or something wrong with it?

**Comment by cebe** [comment]
Finally found the time to review this, merged, Thank you @valinurovam!

**Pull request closed without merging**